Mathlib/CategoryTheory/Bicategory/IsKan.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2023 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import Mathlib.CategoryTheory.Bicategory.Extension
+
+/-!
+# Kan extensions and Kan lifts in bicategories
+
+The left Kan extension of a 1-morphism `g : a ⟶ c` along a 1-morphism `f : a ⟶ b` is the initial
+object in the category of left extensions `LeftExtension f g`.
+
+We introduce a structure `LeftExtension.IsKan t` for an extension `t : LeftExtension f g`. This
+structure consists of a family of 2-morphisms out of `t` together with the fact that such a family
+of 2-morphisms is unique. We have the following definition and lemmas:
+* `LeftExtension.IsKan.desc`: the family of 2-morphisms out of the left Kan extension.
+* `LeftExtension.IsKan.fac`: the unit of any left extension factors through the left Kan extension.
+* `LeftExtension.IsKan.hom_ext`: two 2-morphisms out of a left Kan extension are equal if their
+  compositions with each unit are equal.
+
+## Implementation Notes
+We use the Is-Has design pattern, which is used for the implementation of limits and colimits in
+the category theory library. This means that `is_Kan t` is a structure containing the data of
+2-morphisms which ensure that `t` is a Kan extension, while `HasKan f g` (to be implemented in
+the near future) is a `Prop`-valued typeclass asserting that a Kan extension of `g` along `f`
+exists.
+
+## References
+https://ncatlab.org/nlab/show/Kan+extension
+
+## Todo
+left Kan lifts, right Kan extensions, and right Kan lifts
+-/
+
+namespace CategoryTheory
+
+namespace Bicategory
+
+universe w v u
+
+variable {B : Type u} [Bicategory.{w, v} B] {a b c : B}
+
+namespace LeftExtension
+
+variable {f : a ⟶ b} {g : a ⟶ c}
+
+/-- A left Kan extension of `g` along `f` is an initial object in `LeftExtension f g`. -/
+structure IsKan (t : LeftExtension f g) where
+  /-- The family of morphisms out of a left Kan extension. -/
+  descHom : ∀ s : LeftExtension f g, t ⟶ s
+  /-- The uniqueness of the family of morphisms out of a left Kan extension. -/
+  uniq : ∀ (s : LeftExtension f g) (τ : t ⟶ s), τ = descHom s
```

Comment:
With suitable imports, I would suggest using the existing limits API:
```suggestion
structure IsKan (t : LeftExtension f g) where
  /-- a left Kan extension is initial in `LeftExtension f g` -/
  isInitial : Limits.IsInitial t
```

Reply:
How about stating the equivalence to `IsInitial` as a lemma instead of putting in the default consructor?
The constructor of `IsInitial` (defined as a specific `IsColimit`) unfolds `descHom` into `desc` and `fac`, but I find in the current situation it's more convenient to bundle them when providing a term of `IsKan t`.

Reply:
As it seems to me that your definitions/lemmas are not at all specific to the category `LeftExtension`, I think the overall situation would be better if the general `IsInitial` API was improved instead, and then `IsKan` could be made an abbreviation of `IsInitial`. Indeed, I believe that for the future development of Kan extensions, it shall be convenient to use the `Limits` API, e.g. when two objects in `LeftExtension` are isomorphic, then one is initial iff the other is, etc. This should avoid some unnecessary duplication of code.

Reply:
I see. Maybe it is enough to just define `abbrev IsKan (t : LeftExtension f g) := Limits.IsInitial t`?  I will try it.